ktor-server/ktor-server-netty/jvm/src/io/ktor/server/netty/cio/NettyHttpResponsePipeline.kt

```diff
@@ -0,0 +1,345 @@
+/*
+* Copyright 2014-2021 JetBrains s.r.o and contributors. Use of this source code is governed by the Apache 2.0 license.
+*/
+
+package io.ktor.server.netty.cio
+
+import io.ktor.http.*
+import io.ktor.server.netty.*
+import io.ktor.server.netty.http1.*
+import io.ktor.util.*
+import io.ktor.util.cio.*
+import io.ktor.utils.io.*
+import io.netty.channel.*
+import io.netty.handler.codec.http.*
+import io.netty.handler.codec.http2.*
+import kotlinx.atomicfu.*
+import kotlinx.coroutines.*
+import java.io.*
+import kotlin.coroutines.*
+
+private const val UNFLUSHED_LIMIT = 65536
+
+/**
+ * Contains methods for handling http request with Netty
+ * @param context
+ * @param coroutineContext
+ * @param activeRequests
+ * @param isCurrentRequestFullyRead
+ * @param isChannelReadCompleted
+ */
+@OptIn(InternalAPI::class)
+internal class NettyHttpResponsePipeline constructor(
+    private val context: ChannelHandlerContext,
+    private val httpHandler: NettyHttp1Handler,
+    override val coroutineContext: CoroutineContext
+) : CoroutineScope {
+    /**
+     * True if there is unflushed written data in channel
+     */
+    private val isDataNotFlushed: AtomicBoolean = atomic(false)
+
+    /**
+     * Represents promise which is marked as success when the last read request is handled.
+     * Marked as fail when last read request is failed.
+     * Default value is success on purpose to start first request handle
+     */
+    private var previousCallHandled: ChannelPromise = context.newPromise().also {
+        it.setSuccess()
+    }
+
+    /**
+     * Flush if there is some unflushed data, nothing to read from channel and no active requests
+     */
+    internal fun flushIfNeeded() {
+        if (
+            isDataNotFlushed.value &&
+            httpHandler.isChannelReadCompleted.value &&
+            httpHandler.activeRequests.value == 0L
+        ) {
+            context.flush()
+            isDataNotFlushed.compareAndSet(expect = true, update = false)
+        }
+    }
+
+    internal fun processResponse(call: NettyApplicationCall) {
+        call.previousCallFinished = previousCallHandled
+        call.callFinished = context.newPromise()
+        previousCallHandled = call.callFinished
+
+        processElement(call)
+    }
+
+    private fun processElement(call: NettyApplicationCall) = processOrderedCall(call) {
+        try {
+            processCall(call)
+        } catch (actualException: Throwable) {
+            processCallFailed(call, actualException)
+        } finally {
+            call.responseWriteJob.cancel()
+        }
+    }
+
+    /**
+     * Process [call] with [block] when the response is ready and previous call is successfully processed.
+     * [call] won't be processed with [block] if previous call is failed.
+     */
+    private fun processOrderedCall(call: NettyApplicationCall, block: () -> Unit) {
```

Comment:
The method is setting response ready handler. It should be named somehow like `setOnResponseReadyHandler`. Could you tell me if you have any ideas?